src/hotspot/share/services/nmtPreInit.cpp

```diff
@@ -41,11 +41,22 @@ static void  raw_free(void* p)                  { ALLOW_C_FUNCTION(::free, ::fre
 static const size_t malloc_alignment = 2 * sizeof(void*); // could we use max_align_t?
 STATIC_ASSERT(is_aligned(sizeof(NMTPreInitAllocation), malloc_alignment));
 
+// To keep matters simple we just raise a fatal error on OOM. Since preinit allocation
+// is just used for pre-VM-initialization mallocs, none of which are optional, we don't
+// need a finer grained error handling.
+static void fail_oom(size_t size) {
+  vm_exit_out_of_memory(size, OOM_MALLOC_ERROR, "VM early initialization phase");
+}
+
 // --------- NMTPreInitAllocation --------------
 
 NMTPreInitAllocation* NMTPreInitAllocation::do_alloc(size_t payload_size) {
   const size_t outer_size = sizeof(NMTPreInitAllocation) + payload_size;
+  guarantee(outer_size > payload_size, "Overflow");
```

Comment:
I was suggesting an assertion as this is a VM programming error we should catch during testing in't it?

Reply:
Guarantee as a protection against future bitrot. I am not even 100% sure there are not paths today that are controlled via outside input. I am even less sure about the future. Therefore I protect the release version too. Are you concerned about runtime costs? It is just a comparison.

OTOH I have no strong emotions here. If you insist, I can make it debug only.

Reply:
I'm more concerned that we don't know if this path may be influenced by external input. :(

These kind of "just in case" checks all add up.

But I won't insist on a change.